### PR TITLE
interpreter fundamentals

### PR DESCRIPTION
**Pull request by qwe-q**
Notably, this isn't ready for actual usage yet, because the program cannot load DAP instructions in a loop, but other than that I think the code is fairly sound by now

### CODE REVIEW

File: name-emu/Cargo.toml [hunk 1]
```diff
@@ -6,3 +6,7 @@ edition = "2021"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
+thiserror = "1.0.48"
+dap = "0.3.1-alpha1"
```
Comment (CharlesAverill) [inline name-emu/Cargo.toml:10]: Is the alpha branch safe to use or should we stick with something more stable?

File: name-emu/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+  #[error("Missing command")]
+  MissingCommandError,
+
+  #[error("Command argument error")]
+  CommandArgumentError
+}
+
+type DynResult<T> = std::result::Result<T, Box<dyn std::error::Error>>;
+
+// fn test_mips() {
```
Comment (CharlesAverill) [inline name-emu/src/main.rs:30]: Remove comment or put into mips.rs

File: name-emu/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    }
+
+    // Launch does nothing in NAME, since all state was already set up by the time the protocol reached this point.
+    Command::Launch(_) => {
+
+    }
+
+    Command::WriteMemory(write_mem_args) => {
+      let bytes = general_purpose::STANDARD.decode(write_mem_args.data)?;
+      // let mut i = 0;
```
Comment (CharlesAverill) [inline name-emu/src/main.rs:141]: Remove commented code

File: name-emu/src/mips.rs [hunk 4]
```diff
@@ -0,0 +1,371 @@
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use std::fmt;
+use std::io::Cursor;
+
+pub const DOT_TEXT: u32 = 0x00400000;
```
Comment (CharlesAverill) [inline name-emu/src/mips.rs:5]: Maybe `DOT_TEXT_START_ADDRESS`?

Comment (CharlesAverill) [reply name-emu/src/mips.rs:5]: or instead of `DOT_TEXT...`, `TEXT_SECTION_....`

File: name-emu/src/mips.rs [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+            if let Some(value) = memory.get(offset as usize) {
+                Ok(*value)
+            }
+            else {
+                Err(ExecutionErrors::MemoryObviouslyUninitializedAccess)
+            }
+        }
+        else { Err(ExecutionErrors::MemoryIllegalAccess) }
+    }
+    pub fn read_h(&mut self, address: u32) -> Result<u16, ExecutionErrors> {
```
Comment (CharlesAverill) [inline name-emu/src/mips.rs:306]: `read_halfword`?

File: name-emu/src/mips.rs [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+        for (pool, base_address, max_length) in &mut self.memories {
+            if (*base_address .. *base_address + *max_length).contains(&address) {
+                return Some((pool, address - *base_address))
+            }
+        }
+        None
+    }
+
+    // This function attempts to access memory and returns an error if that memory doesn't exist
+    pub fn read_b(&mut self, address: u32) -> Result<u8, ExecutionErrors> {
```
Comment (CharlesAverill) [inline name-emu/src/mips.rs:295]: `read_byte`?

File: name-emu/src/mips.rs [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+                Err(ExecutionErrors::MemoryObviouslyUninitializedAccess)
+            }
+        }
+        else { Err(ExecutionErrors::MemoryIllegalAccess) }
+    }
+    pub fn read_h(&mut self, address: u32) -> Result<u16, ExecutionErrors> {
+        let bytes = [self.read_b(address)?, self.read_b(address + 1)?];
+        Ok(Cursor::new(bytes).read_u16::<LittleEndian>().unwrap())
+    }
+    pub fn read_w(&mut self, address: u32) -> Result<u32, ExecutionErrors> {
```
Comment (CharlesAverill) [inline name-emu/src/mips.rs:310]: `read_word`?

File: name-emu/src/mips.rs [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+        let bytes = [self.read_b(address)?, self.read_b(address + 1)?,
+                        self.read_b(address + 2)?, self.read_b(address + 3)?];
+        Ok(Cursor::new(bytes).read_u32::<LittleEndian>().unwrap())
+    }
+
+    
+    // I'm clueless on how to expand this. I need to talk to Cole
+    pub fn write_b(&mut self, address: u32, value: u8) -> Result<(), ExecutionErrors> {
```
Comment (CharlesAverill) [inline name-emu/src/mips.rs:318]: Same comments as read function naming

File: name-emu/Cargo.toml (hunk 1 above)
Comment (qwe-q) [reply name-emu/Cargo.toml:10]: Unfortunately dap-rs does not yet have a stable version. We will have to live with this and keep it up to date

File: name-emu/src/mips.rs (hunk 4 above)
Comment (qwe-q) [reply name-emu/src/mips.rs]: done

**Review by CharlesAverill — APPROVED** [review]

**Review by CowTipper989 — APPROVED** [review]

**Review by hkwheat — APPROVED** [review]

**Pull request merged**